### ISSUE TITLE
proposal: `unsafe_future_or_is_check`

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#4818]
When using generic `FutureOr`s, a possible mistake is mishandling `FutureOr<Object?>`.

Consider:

```dart
void fn<T>(FutureOr<T> futureOr) {
  if (futureOr is T) {
    print(value);
  } else {
    print(future);
  }
}
```

One issue with this check is that in the case of `T == Object`, the first `if` is always true.
Meaning that when doing `fn<Object>(Future.value(...))`, this will print "value" instead of "future".

## Propopal

We could have a lint that warns when doing `futureOr is T` where `T` is a generic that allows  `Future<T>`

```dart
void fn<T>(FutureOr<T> futureOr) {
  if (futureOr is T) {} // KO

  if (futureOr is Future<T>) {} // OK
}

void fn<T extends num>(FutureOr<T> futureOr) {
  if (futureOr is T) {} // OK, "num" cannot be a Future
}

void fn<T extends Object>(FutureOr<T> futureOr) {
  if (futureOr is T) {} // KO, T can be a Future<Object>
}
```

**Comment by srawlins** [comment]
I support. Great catch; it took me a second!

**Comment by rrousselGit** [comment]
This was learned the hard way 😛

**Comment by eernstg** [comment]
One way to make this safer is to reorder:

```dart
import 'dart:async';

void fn<T>(FutureOr<T> x) async {
  if (x is Future<T>) {
    print(await x);
  } else {
    print(x);
  }
}

void main() {
  fn<Object>(Future<Object?>.value(null)); // "Instance of '_Future<Object?>'".
}
```

An interesting case is the case where a given future can be considered as belonging to the both branches of the union type, that is, `FutureOr<T>` for a `T` which is `Object` or a supertype thereof. But also the case where a future can _only_ be considered as belonging to the non-future branch, like the example above.

In this case we are invoking `fn` with a `Future<Object?>` which must be interpreted as an `Object` in order to be a value whose type can be `FutureOr<Object>`. So in that sense it isn't a future "of the right kind" (in particular, it can be completed with `null`, so it can't possibly be typed as a `Future<Object>`). Hence, it isn't awaited in `fn`, it is just printed as it is. If the type `FutureOr<T>` is to be taken seriously then I think that's the most sane behavior.

Here is an example where the given object belongs to both branches, illustrating that it doesn't only happen with `Object`:

```dart
import 'dart:async';

class C {} // Could have long paths to object (`implements ...`).

class D implements Future<C>, C {
  final Future<C> future = Future.value(C());
  D();
  @override asStream() => future.asStream();
  @override catchError(f, {test}) => future.catchError(f, test: test);
  @override then<R>(f, {Function? onError}) => future.then(f, onError: onError);
  @override timeout(d, {onTimeout}) => future.timeout(d, onTimeout: onTimeout);
  @override whenComplete(f) => future.whenComplete(f);
}

void fn<T>(FutureOr<T> x) async {
  if (x is Future<T>) {
    print(await x);
  } else {
    print(x);
  }
}

void main() {
  fn<C>(D()); // "Instance of 'C'".
}
```

In this case it also makes a difference whether we're testing for `T` or `Future<T>`. It isn't easy to say what the "correct" choice is, but it seems likely that we'd like to await a future if we have one. So, again, we should test for `Future<T>`.

Maybe the lint should simply say **"Consider testing `o is Future<T>` rather than `o is T`"** in the situation where the static type of `o` is `FutureOr<T>`. We might want to avoid flagging cases where the target object cannot possibly belong to both branches, but we actually don't have absolute certainty about that very often. For instance, we could write something like class `D` above with any `C` that doesn't have a member name conflict with `Future`.

**Comment by lrhn** [comment]
I've also learned to always do the `is Future<X>` check, never the `is X`. I don't even think about it any more, it's just reflex.
Definitely something to warn about.

This could be a "potentially dead code" warning, rather than being specific to `FutureOr`.

The same thing  *could* apply to `T? x = ...; if (x is T) { ... } else { ... }`, but there are cases where I actually want to treat `null` as `T` when `T` is nullable, and cases where I want to treat it as `null`, so it's not as clear-cut.
Futures are something you never want to ignore and treat as a non-future.
